client/fingerprint/env_aws.go

```diff
@@ -104,24 +108,25 @@ func (f *EnvAWSFingerprint) Fingerprint(request *FingerprintRequest, response *F
 	}
 
 	for k, unique := range keys {
-		resp, err := ec2meta.GetMetadata(k)
-		v := strings.TrimSpace(resp)
-		if v == "" {
-			f.logger.Debug("read an empty value", "attribute", k)
-			continue
-		} else if awsErr, ok := err.(awserr.RequestFailure); ok {
-			f.logger.Debug("could not read attribute value", "attribute", k, "error", awsErr)
+		resp, err := imdsClient.GetMetadata(ctx, &imds.GetMetadataInput{
+			Path: k,
+		})
+		if err := f.handleImdsError(err, k); err != nil {
+			return err
+		}
+		if resp == nil {
 			continue
-		} else if awsErr, ok := err.(awserr.Error); ok {
-			// if it's a URL error, assume we're not in an AWS environment
-			// TODO: better way to detect AWS? Check xen virtualization?
-			if _, ok := awsErr.OrigErr().(*url.Error); ok {
-				return nil
-			}
+		}
+		defer resp.Content.Close()
 
-			// not sure what other errors it would return
+		bytes, err := io.ReadAll(resp.Content)
+		if err != nil {
 			return err
 		}
+		v := strings.TrimSpace(string(bytes))
+		if v == "" {
+			f.logger.Debug("read an empty value", "attribute", k)
+		}
```

Comment:
Do we want a continue statement after the log line here which will match the existing behaviour and mean we do not have empty string attributes in node data? It would also be nice if we do need this continue to craft a test to catch this in the future, although I don't know if that is possible or feasible.

Reply:
In my testing, I haven't seen IMDS return empty data.  If an attribute wasn't available, the actual endpoint is not available and we get a 404 which is handled via `handleImdsError()`.

There is currently a test that returns empty data and mentions catching this in Hetzner cloud, but it's a false positive because it doesn't include the `ami-id`, so the code will actually return earlier in the fingerprint.  Also Hetzner uses a different metadata path (but maybe they didn't 4 years ago when this test was written).

I'd say we keep it as a "just-in-case" and I can update the test/comments accordingly?